### ISSUE TITLE
Missing laws

### ISSUE THREAD

**Issue by ekmett** [ekmett/mtl#5]
The laws for `MonadState`, etc. are missing from the haddocks.

It may be a good idea to add them.

**Comment by Lysxia** [comment]
@ekmett That paper states `MonadState` laws in terms of `get` and `modify` (called `update` (`u`) in the paper).

```haskell
modify f >> modify g   =   modify (g . f)  -- modify-modify
get >>= \s1 -> get -> \s2 -> return (s1, s2)   =   get >>= \s -> return (s, s)   -- get-get
get >>= \s1 -> modify f >> get >>= \s2 -> return (s1, s2)   =   get >>= \s -> modify f >> return (s, f s)   -- get-modify-get
```

Since mtl's `MonadState` is given by `get` and `put`, it seems simpler to reformulate laws in these terms. The four laws obtained from the different ways of sequencing two of `put` or `get` imply the above, but I'm wondering whether these are too strong for some useful instances, where `get` and `put` may have more side-effects besides querying/modifying the state.

```haskell
get >>= \s1 -> get -> \s2 -> return (s1, s2)   =   get >>= \s -> return (s, s)   -- get-get (the same)
get >>= \_ -> put s   =   put s   -- get-put
put s1 >> put s2   =   put s2   -- put-put
put s >> get    =   put s >> return s   -- put-get
```

The following `get-put-2` seems a bit more "economical" by making use of just two class method calls instead of three in `get-put`; together with `get-get`, it implies `get-put`.

```haskell
get >>= \s -> put s   =   return ()   -- get-put-2
```

In a similar vein, one might think of `get-void` below:

```haskell
void get = return ()   -- get-void
```

Are the equations above general enough? Is there a litmus test for sufficiently general `MonadState` laws (and for other mtl classes)?

This may just be my wishful thinking, but one nice-sounding property is that any sequence of `get` and `put` can be rewritten as a single `get` followed by a single `put`, and maybe a `return`: `get s >>= put (f s) >> return (g s)`. To that end, `get-get`, `put-put` and `put-get` take care of most cases, then `get-put-2` seems just right to finish that "normalization".

**Comment by Lysxia** [comment]
For `MonadReader`, may I suggest the following:

```haskell
(,) <$> ask <*> ask   =   (\r -> (r, r)) <$> ask   -- ask-ask
local f ask   =   fmap f ask   -- local-ask
local f . local g   =   local (g . f)   -- local-local

local f (return x)   =   return x   -- local-M-return
local f (m >>= \x -> k x)   =   local f m >>= \x -> local f (k x)   -- local-M-bind
```

**Comment by ocharles** [comment]
Could your `ask-ask` law just be

```haskell
ask >> ask = ask
```

?

**Comment by Lysxia** [comment]
@ocharles Right, that looks equivalent and simpler. Formally speaking, I'm missing a step to prove that `ask >> ask = ask` implies `(,) <$> ask <*> ask = (\r -> (r, r)) <$> ask`; does it need some extra assumption about the monad (that might always hold in the cases we care about)?

**Comment by ekmett** [comment]
You could probably go even stronger with something like

```haskell
    m <*> ask = ask <**> m
    ask *> m = m = m <* ask
```

to show that `ask` commutes with everything, has no other side-effects and no monadic action can affect the answer of `ask`.

**Comment by ekmett** [comment]
Also: at the very least, `reader`, `writer` and `state` are all monad homomorphisms, just like `lift`. (They are even almost always monad monomorphisms, but for an annoying technical corner case involving the terminal monad transformer.)

**Comment by Lysxia** [comment]
Interesting, these homomorphisms imply a lot of the laws.

For example, *obviously* in `Reader` we can check that `ask >> ask = ask`. Now in a general `MonadReader`, provided `ask = lift ask`, we have `ask >> ask = lift ask >> lift ask = lift (ask >> ask) = lift ask = ask`.

**Comment by recursion-ninja** [comment]
Any progress on finalizing these laws and adding them to the documentation?

**Comment by Lysxia** [comment]
Thanks for the reminder @recursion-ninja . I think this issue was a victim of "perfect is the enemy of good". There are so many ways to present the laws, and comparing them all against each other is a deep rabbit hole. I just opened PR #61 to get things started.

**Comment by Zemyla** [comment]
The `get`-`put` laws can be easier described if they're defined in terms of `state`, I think.

```haskell
-- Laws for get and put
get = state $ \s -> (s, s)
put s = state $ const $ ((), s)

-- Law for state joining
state f >>= state . g = state $ uncurry g . f
```

Also, I'm wondering whether `MonadCont` has any laws. From what I can tell, these sound reasonable:

```haskell
-- These laws specify the basic behavior of the passed continuation.

-- callCC itself has no side effects.
callCC (const m) = m

-- The continuation passed returns the given value.
callCC ($ a) = pure a

-- It also does so in the presence of side effects.
callCC ((>>=) m) = m

-- The return value of the continuation passed is actually Void, so it doesn't "really" return.
callCC f = callCC (f . fmap absurd)

-- And then this law I think would allow decomposition of the behavior of callCC by induction into the base cases. Though I'm not sure if it actually works that way.
callCC (\k -> f k >>= (\a -> k a >>= g k)) = callCC f
```

I'm also not sure whether the situation with `StateT` on `MonadCont` would screw things up.

**Comment by Lysxia** [comment]
@Zemyla Those laws look great!

Here they are QuickChecked: https://github.com/Lysxia/checkers-mtl/blob/with-higherorder/src/Test/Monad/Cont.hs

The tests pass with those monads at least: https://github.com/Lysxia/checkers-mtl/blob/d30c35ea87d1746bbd2d1728abd0045f8355b650/src/Test/Monad/Cont/Checkers.hs#L36-L39

**Comment by chessai** [comment]
Keep forgetting to note here, that as soon as the laws are decided on and a PR is merged, i will add them to quickcheck-classes and hedgehog-classes.

**Comment by treeowl** [comment]
It'd be great to actually get this sorted. The control operations are a horrible mess for things like `list-t` and `logict`, and not even knowing how they're *supposed* to behave makes implementation practically impossible.
